File: src-tauri/src/commands/sandbox.rs
//! Tauri commands for the three-tier sandbox.
//!
//! The runtime gate lives in `chat.rs`; these commands only let the UI read
//! and persist the configured tier per project.

use std::path::PathBuf;

use crate::orchestrator::{is_read_only_command, load_tier, write_tier, SandboxTier};

/// Result of classifying a shell command for the UI (e.g. the `/run` composer
/// or an approval prompt): is it provably read-only, with a one-line reason?
#[derive(Debug, serde::Serialize)]
pub struct CommandClassification {
    pub read_only: bool,
    pub reason: String,
}

/// Codex-style classification of a shell command line. A `read_only:true`
/// command is safe to run under any sandbox tier (it cannot mutate the
/// filesystem, network, or process state); anything else requires
/// `workspace-write`/`danger-full-access` or an explicit approval.
#[tauri::command]
pub async fn classify_shell_command(command: String) -> CommandClassification {
    let read_only = is_read_only_command(&command);
    let reason = if read_only {
        "read-only — inspects only; runs under any sandbox tier".to_string()
    } else {
        "not classified read-only — may modify the workspace; needs workspace-write or approval"
            .to_string()
    };
    CommandClassification { read_only, reason }
}

/// Read the configured sandbox tier for `project_root`. Missing /
/// malformed files yield the default (`workspace-write`).
#[tauri::command]
pub async fn get_sandbox_tier(project_root: String) -> Result<String, String> {
    if project_root.trim().is_empty() {
        return Err("project_root is required".into());
    }
    let root = PathBuf::from(&project_root);
    Ok(load_tier(&root).as_str().to_string())
}

/// Persist the sandbox tier for `project_root`. Accepted values:
/// `"read-only"`, `"workspace-write"`, `"danger-full-access"` (plus snake/
/// camel variants — see `SandboxTier::parse`).
#[tauri::command]
pub async fn set_sandbox_tier(project_root: String, tier: String) -> Result<(), String> {
    if project_root.trim().is_empty() {
        return Err("project_root is required".into());
    }
    let parsed = SandboxTier::parse(&tier)
        .ok_or_else(|| format!("invalid tier '{tier}' (use read-only|workspace-write|danger-full-access)"))?;
    let root = PathBuf::from(&project_root);
    write_tier(&root, parsed).map_err(|e| e.to_string())
}

File: src/lib/sandbox.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Codex-style three-tier sandbox. Mirrors `SandboxTier` in
 * `src-tauri/src/orchestrator/sandbox.rs`.
 *
 *  - `read-only`          — only read-shaped tools (read/search/ls/grep/…).
 *  - `workspace-write`    — read + write/edit/patch/run_*, writes must stay
 *                           inside the project root.
 *  - `danger-full-access` — anything goes; no gate.
 */
export type SandboxTier = "read-only" | "workspace-write" | "danger-full-access";

export const SANDBOX_TIERS: readonly SandboxTier[] = [
  "read-only",
  "workspace-write",
  "danger-full-access",
] as const;

/** Default matches the Rust default — historic behavior is workspace-write. */
export const DEFAULT_SANDBOX_TIER: SandboxTier = "workspace-write";

function isTier(s: string): s is SandboxTier {
  return (SANDBOX_TIERS as readonly string[]).includes(s);
}

/** Read the configured sandbox tier for `projectRoot`. Returns the default
 *  when no `.cortex/sandbox.toml` exists or it fails to parse. */
export async function getSandboxTier(projectRoot: string): Promise<SandboxTier> {
  const raw = await invoke<string>("get_sandbox_tier", { projectRoot });
  return isTier(raw) ? raw : DEFAULT_SANDBOX_TIER;
}

/** Persist the sandbox tier for `projectRoot`. Rejects on invalid tier
 *  strings; the Rust side does its own validation too. */
export async function setSandboxTier(
  projectRoot: string,
  tier: SandboxTier,
): Promise<void> {
  if (!isTier(tier)) {
    throw new Error(`invalid sandbox tier '${tier}'`);
  }
  return invoke("set_sandbox_tier", { projectRoot, tier });
}

/** Classification of a shell command line. Mirrors `CommandClassification`
 *  in `src-tauri/src/commands/sandbox.rs`. A `readOnly:true` command (Codex's
 *  `is_safe_command`) is safe to run under any tier — including `read-only`,
 *  which is what an untrusted project is forced into. */
export interface CommandClassification {
  readOnly: boolean;
  reason: string;
}

/** Ask the backend whether a shell command is provably read-only (safe to run
 *  under the read-only sandbox tier). Use to label `/run` input or an approval
 *  prompt. */
export async function classifyShellCommand(
  command: string,
): Promise<CommandClassification> {
  const r = await invoke<{ read_only: boolean; reason: string }>(
    "classify_shell_command",
    { command },
  );
  return { readOnly: r.read_only, reason: r.reason };
}

/** Human label + short description, for use in popovers and settings. */
export const SANDBOX_TIER_META: Record<
  SandboxTier,
  { label: string; description: string; color: string }
> = {
  "read-only": {
    label: "Read-only",
    description:
      "Read tools + provably read-only shell commands (git status, ls, grep…).",
    color: "var(--sandbox-readonly, #3b82f6)",
  },
  "workspace-write": {
    label: "Workspace write",
    description: "Read + write/edit/patch/run inside the project root.",
    color: "var(--sandbox-workspace, #22c55e)",
  },
  "danger-full-access": {
    label: "Danger: full access",
    description: "No tier gate. Guardrails still apply.",
    color: "var(--sandbox-danger, #ff6b6b)",
  },
};
